File: android/app/src/main/java/dev/walnuts/beast/domain/model/Video.kt
package dev.walnuts.beast.domain.model

import kotlinx.serialization.Serializable

@Serializable
enum class VideoStatus { UPLOADED, ENCODING, READY, FAILED }

@Serializable
enum class PlaybackSource { ENCRYPTED_DASH, DEBUG_PREVIEW }

@Serializable
data class EncryptionMetadata(
    val algorithm: String,
    val chunkSize: Int,
    val keyVersion: String,
    val nonce: String,
    val encryptedDataKey: String,
    val sharedKeyId: String,
)

@Serializable
data class Video(
    val id: String,
    val status: VideoStatus,
    val tags: List<String>,
    val playCount: Int,
    val rating: Int?,
    val lastPlayedAt: String?,
    val encryption: EncryptionMetadata,
    val encryptedTags: String? = null,
    val thumbnailUrl: String? = null,
    val playbackSource: PlaybackSource = PlaybackSource.ENCRYPTED_DASH,
    val encryptedDashManifestUrl: String? = null,
    val playbackUrl: String? = null,
    val durationSeconds: Long? = null,
    val progress: Float = if (status == VideoStatus.READY) 1f else 0f,
)

enum class RecommendationKind(val label: String, val description: String) {
    FAVORITES("あなたのお気に入り", "高く評価した動画"),
    RETURN_TO_FAVORITES("最近再生していないお気に入り", "少し時間が空いた動画"),
    UNWATCHED("まだ再生していない動画", "新しい動画を見つける"),
    RECENT("最近見た動画", "続きを楽しむ"),
}

fun Video.isFavorite() = rating != null && rating >= 4

fun Video.recommendationKinds(nowEpochSeconds: Long): Set<RecommendationKind> = buildSet {
    if (isFavorite()) add(RecommendationKind.FAVORITES)
    val lastPlayedEpoch = lastPlayedAt?.toLongOrNull()
    if (isFavorite() && (lastPlayedEpoch == null || nowEpochSeconds - lastPlayedEpoch > 60 * 60 * 24 * 14)) {
        add(RecommendationKind.RETURN_TO_FAVORITES)
    }
    if (playCount == 0) add(RecommendationKind.UNWATCHED)
    if (lastPlayedEpoch != null && nowEpochSeconds - lastPlayedEpoch <= 60 * 60 * 24 * 7) {
        add(RecommendationKind.RECENT)
    }
}

File: backend/internal/domain/video.go
package domain

import (
	"errors"
	"time"

	"uuid"
)

var (
	ErrVideoNotFound = errors.New("video not found")
	ErrInvalidRating = errors.New("rating must be between 1 and 5")
	ErrInvalidStatus = errors.New("invalid video status")
)

type VideoStatus string

const (
	VideoStatusUploaded VideoStatus = "UPLOADED"
	VideoStatusEncoding VideoStatus = "ENCODING"
	VideoStatusReady    VideoStatus = "READY"
	VideoStatusFailed   VideoStatus = "FAILED"
)

type EncryptionMetadata struct {
	Algorithm        string
	ChunkSize        int
	KeyVersion       string
	Nonce            string
	EncryptedDataKey string
	SharedKeyID      string
}

type DashArtifact struct {
	ObjectKey  string
	Encryption EncryptionMetadata
}

func (v *Video) TransitionStatus(next VideoStatus) error {
	valid := map[VideoStatus]bool{VideoStatusUploaded: true, VideoStatusEncoding: true, VideoStatusReady: true, VideoStatusFailed: true}
	if !valid[next] {
		return ErrInvalidStatus
	}
	allowed := map[VideoStatus]map[VideoStatus]bool{
		VideoStatusUploaded: {VideoStatusEncoding: true},
		VideoStatusEncoding: {VideoStatusReady: true, VideoStatusFailed: true},
		VideoStatusFailed:   {VideoStatusEncoding: true},
		VideoStatusReady:    {},
	}
	if !allowed[v.Status][next] {
		return ErrInvalidStatus
	}
	v.Status = next
	v.UpdatedAt = time.Now().UTC()
	return nil
}

type Video struct {
	ID              string
	OwnerID         string
	Status          VideoStatus
	ObjectKey       string
	SourceObjectKey string
	EncryptedTags   string
	PlayCount       int64
	Rating          *int
	LastPlayedAt    *time.Time
	CreatedAt       time.Time
	UpdatedAt       time.Time
	Encryption      EncryptionMetadata
	Progress        float64
	ErrorMessage    string
	DashArtifacts   map[string]DashArtifact
}

func NewVideo(ownerID, encryptedTags, objectKey string, encryption EncryptionMetadata) (Video, error) {
	if ownerID == "" || encryptedTags == "" || objectKey == "" || encryption.SharedKeyID == "" {
		return Video{}, errors.New("owner, encrypted tags, object key, and shared key are required")
	}
	now := time.Now().UTC()
	return Video{
		ID:              uuid.New().String(),
		OwnerID:         ownerID,
		Status:          VideoStatusUploaded,
		ObjectKey:       objectKey,
		SourceObjectKey: objectKey,
		EncryptedTags:   encryptedTags,
		CreatedAt:       now,
		UpdatedAt:       now,
		Encryption:      encryption,
		DashArtifacts:   make(map[string]DashArtifact),
	}, nil
}

func (v *Video) SetProgress(progress float64) error {
	if progress < 0 || progress > 1 {
		return errors.New("progress must be between 0 and 1")
	}
	v.Progress = progress
	v.UpdatedAt = time.Now().UTC()
	return nil
}

func (v *Video) SetRating(rating *int) error {
	if rating != nil && (*rating < 1 || *rating > 5) {
		return ErrInvalidRating
	}
	v.Rating = rating
	v.UpdatedAt = time.Now().UTC()
	return nil
}

func (v *Video) RecordPlayback(now time.Time) {
	v.PlayCount++
	t := now.UTC()
	v.LastPlayedAt = &t
	v.UpdatedAt = t
}

func (v *Video) SetStatus(status VideoStatus) error {
	return v.TransitionStatus(status)
}
